xayn-ai/src/coi/system.rs

```diff
@@ -399,48 +443,92 @@ impl CoiSystem {
                 (relevance > 0.)
-                    .then(|| (relevance / max).clamp(0., 1.))
+                    .then(|| Relevance((relevance / max).max(0.).min(1.)))
                     .unwrap_or_default()
             });
 
-            let key_phrases = selected
+            for (key_phrase, relevance) in selected
                 .iter()
-                .zip(
-                    coi.key_phrases()
-                        .iter()
-                        .map(Cow::Borrowed)
-                        .chain(candidates.into_iter().map(Cow::Owned)),
-                )
-                .filter_map(|(is_selected, key_phrase)| {
-                    is_selected.then(|| key_phrase.into_owned())
-                })
-                .zip(relevance)
-                .filter_map(|(key_phrase, relevance)| key_phrase.with_relevance(relevance).ok())
-                .collect();
-            coi.set_key_phrases(key_phrases);
+                .zip(key_phrases)
+                .filter_map(|(is_selected, key_phrase)| is_selected.then(|| key_phrase))
+                .zip(relevances)
+            {
+                system
+                    .coi_to_relevance
+                    .entry(coi_id)
+                    .or_default()
+                    .insert(relevance);
+                system
+                    .relevance_to_key_phrase
+                    .entry((relevance, coi_id))
+                    .or_default()
+                    .push(key_phrase);
+            }
         }
 
-        let candidates = unique_candidates(coi, candidates, &self.smbert);
-        let (similarity, normalized) = similarities(coi, &candidates);
-        let selected = is_selected(normalized, self.config.max_key_phrases, self.config.gamma);
-        select(coi, candidates, selected, similarity);
+        let key_phrases = unique(self, coi.id(), candidates);
+        let (similarity, normalized) = similarities(&key_phrases, coi.point());
+        let selected = is_selected(normalized, self.config.penalty.len(), self.config.gamma);
+        select(self, coi.id(), key_phrases, selected, similarity);
+    }
+
+    /// Selects the top key phrases from the cois, sorted in descending relevance.
+    ///
+    /// The selected key phrases and their relevances are removed from the maps of this system.
+    #[allow(dead_code)]
+    fn select_top_key_phrases<CP: CoiPoint + CoiPointStats>(
+        &mut self,
+        cois: &mut [CP],
+        top: usize,
+    ) -> Vec<KeyPhrase> {
+        self.compute_relevances(cois);
+        // TODO: refactor once pop_last() etc are stabilized for BTreeMap
+        let (relevance_to_coi, key_phrases) = self
+            .relevance_to_key_phrase
+            .iter()
+            .map(|((relevance, coi_id), key_phrases)| {
+                key_phrases
+                    .iter()
+                    .map(move |key_phrase| ((*relevance, *coi_id), key_phrase.clone()))
+            })
+            .flatten()
+            .rev()
```

Comment:
not sure what you mean by before `iter()`, do you mean before the `flatten()` or the `map()`? it might only change the order of key phrases with equal relevance, but we don't guarantee that order anyways, so it should be ok to change it. i'll test it and see if i missed sth.